File: LeetCode_ProblemSet/Problem2999.cpp
#include <cmath>
#include <iostream>
#include <string>

class Solution {
   public:
    long long numberOfPowerfulInt(long long start, long long finish, int limit, std::string s) {
        return (
            count(std::to_string(finish), s, limit) -
            count(std::to_string(start - 1), s, limit));
    }

   private:
    long long count(std::string finish, std::string s, int limit) {
        if (finish.length() < s.length()) return 0;
        if (finish.length() == s.length()) return finish >= s;

        std::string suffix = finish.substr(finish.length() - s.length(), s.length());
        long long ans = 0;
        int prefixLength = finish.length() - s.length();

        for (int digit = 0; digit < prefixLength; digit++) {
            if (limit < (finish[digit] - '0')) return ans += (long long)std::pow(limit + 1, prefixLength - digit);

            ans += (long long)(finish[digit] - '0') * (long long)std::pow(limit + 1, prefixLength - 1 - digit);
        }

        ans += suffix >= s;

        return ans;
    }
};

File: LeetCode_ProblemSet/Rust/src/Problem2999.rs
impl Solution {
    pub fn number_of_powerful_int(start: i64, finish: i64, limit: i32, s: String) -> i64 {
        return Solution::count(&finish.to_string(), &s, limit)
            - Solution::count(&(start - 1).to_string(), &s, limit);
    }

    fn count(finish: &str, s: &str, limit: i32) -> i64 {
        if finish.len() < s.len() {
            return 0;
        }
        if finish.len() == s.len() {
            if finish >= s {
                return 1;
            } else {
                return 0;
            }
        }

        let suffix = &finish[(finish.len() - s.len())..];
        let mut ans = 0;
        let prefix_len = finish.len() - s.len();

        for digit in 0..prefix_len {
            if limit < (finish.chars().nth(digit).unwrap() as i32 - '0' as i32) {
                ans += ((limit + 1) as i64).pow((prefix_len - digit) as u32) as i64;
                return ans;
            }
            ans += (finish.chars().nth(digit).unwrap() as i32 - '0' as i32) as i64 * ((limit + 1) as i64).pow((prefix_len - digit - 1) as u32) as i64;
        }
        if suffix >= s.clone() {
            ans += 1;
        }

        return ans;
    }
}
